### PR TITLE
KAFKA-4786: Wait for heartbeat thread to terminate in consumer close

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@hachikuji @ijuma Can you take a look? Thanks.

At the moment, `consumer.close()` marks heartbeat thread as closed, but doesn't do a join, so the thread may still be alive when `close()` returns. I could update the test to `waitForCondition` and give some time for the thread to terminate, but I changed the implementation to do a `join()` similar to what the producer does when force closing the I/O thread, since it is nice to have everything cleaned up before `close()` returns.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -315,6 +315,19 @@ private synchronized void disableHeartbeatThread() {
...
     }
 
+    private void closeHeartbeatThread() {
+        if (heartbeatThread != null) {
+            heartbeatThread.close();
+
+            try {
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
+                log.error("Interrupted while waiting for consumer heartbeat thread to close");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: Should this be a warning or even `info`? It doesn't seem like there's much for the user to do in this case.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, seems to make sense to me, just left one comment. Let's see what @hachikuji says.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java]: @ijuma Thank you for the review. I have changed to `warn`.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the fix!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 2]
```diff
@@ -315,6 +315,19 @@ private synchronized void disableHeartbeatThread() {
             heartbeatThread.disable();
     }
 
+    private void closeHeartbeatThread() {
+        if (heartbeatThread != null) {
+            heartbeatThread.close();
+
+            try {
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:324]: Actually I'm wondering if we should raise this directly as the kafka `InterruptException` to be consistent with how we handle interrupts in `ConsumerNetworkClient`?

Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:324]: @hachikuji Thank you for the review. Updated to throw `InterruptException`.

**Review by apurvam — APPROVED** [review]
LGTM!

**Comment by hachikuji** [comment]
LGTM (for real this time). Merging to trunk and 0.10.2.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 3]
```diff
@@ -315,6 +315,19 @@ private synchronized void disableHeartbeatThread() {
...
+                log.warn("Interrupted while waiting for consumer heartbeat thread to close");
+                throw new InterruptException(e);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:326]: We should throw the original exception here right?

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:326]: `InterruptedException` is checked, so we can't throw it.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:326]: My bad, I hadn't noticed that the rethrown exception was actually different (subtle name difference). I blame the small screen on my phone. ;)

**Pull request closed without merging**